pkg/runtime/wfengine/activity.go

```diff
@@ -50,15 +72,68 @@ func (a *activityActor) SetActorRuntime(actorsRuntime actors.Actors) {
 // returns immediately after creating the reminder, enabling the workflow to continue processing other events
 // in parallel.
 func (a *activityActor) InvokeMethod(ctx context.Context, actorID string, methodName string, data []byte) (any, error) {
-	// This actor doesn't define any methods, and uses the method name and payload as the reminder name and payload.
-	err := a.createReliableReminder(ctx, actorID, methodName, data)
+	var ar ActivityRequest
+	if err := actors.DecodeInternalActorData(data, &ar); err != nil {
+		return nil, fmt.Errorf("failed to decode activity request: %w", err)
+	}
+
+	// Try to load activity state. If we find any, that means the activity invocation is a duplicate.
+	if state, err := a.loadActivityState(ctx, actorID, ar.Generation); err != nil {
+		return nil, err
+	} else if state.Generation > 0 {
+		return nil, ErrDuplicateInvocation
+	}
+
+	// Save the request details to the state store in case we need it after recovering from a failure.
+	state := activityState{
+		Generation:   ar.Generation,
+		EventPayload: ar.HistoryEvent,
+	}
+	if err := a.saveActivityState(ctx, actorID, state); err != nil {
+		return nil, err
+	}
+
+	// The actual execution is triggered by a reminder
+	err := a.createReliableReminder(ctx, actorID, "run-activity", ar.Generation)
 	return nil, err
 }
 
 // InvokeReminder implements actors.InternalActor and executes the activity logic.
-// TODO: Need to review how errors for reminders are handled.
 func (a *activityActor) InvokeReminder(ctx context.Context, actorID string, reminderName string, data any, dueTime string, period string) error {
-	taskEvent, err := backend.UnmarshalHistoryEvent(data.([]byte))
+	wfLogger.Debugf("invoking reminder '%s' on activity actor '%s'", reminderName, actorID)
```

Comment:
Is there any chance this may include PII? If so, we shouldn't log it

Reply:
No, we control the name of the reminder. However, we only have a single reminder name, so there's probably no benefit in logging it. I can remove that part of this statement.